### PR TITLE
Allow new formatter styles for trailing comma and whitespace around proc literal

### PR DESCRIPTION
**Pull request by Blacksmoke16**
Addresses https://github.com/crystal-lang/crystal/pull/14718#issuecomment-2173800679 by allowing new formatter features but without enforcing them.

Open to suggestions on how to do this better, but this seems to work :shrug:.

### CODE REVIEW

File: spec/compiler/formatter/formatter_spec.cr [hunk 1]
```diff
@@ -1,8 +1,8 @@
 require "spec"
 require "../../../src/compiler/crystal/formatter"
 
-private def assert_format(input, output = input, strict = false, flags = nil, file = __FILE__, line = __LINE__)
-  it "formats #{input.inspect}", file, line do
+private def assert_format(input, output = input, strict = false, flags = nil, file = __FILE__, line = __LINE__, focus = false)
```
Comment (Blacksmoke16) [inline spec/compiler/formatter/formatter_spec.cr:4]: This proved useful when testing, but happen to move it to another PR if you'd rather.

File: spec/compiler/formatter/formatter_spec.cr [hunk 2]
```diff
@@ -812,7 +812,7 @@ describe Crystal::Formatter do
       end
       CRYSTAL
       def foo(x,
-              y)
+              y,)
```
Comment (Blacksmoke16) [inline spec/compiler/formatter/formatter_spec.cr:815]: This one is a bit weird IMO, but formatter with `def_trailing_comma` also produces this output, so prob best to leave it as its own bug/enhancement.

File: src/compiler/crystal/tools/formatter.cr [hunk 3]
```diff
@@ -4272,15 +4273,15 @@ module Crystal
...
+      write " " if a_def.args.present? || return_type || flag?("proc_literal_whitespace") || whitespace_after_op_minus_gt
 
       is_do = false
       if @token.keyword?(:do)
         write_keyword :do
         is_do = true
       else
         write_token :OP_LCURLY
-        write " " if a_def.body.is_a?(Nop) && flag?("proc_literal_whitespace")
+        write " " if (a_def.body.is_a?(Nop) && flag?("proc_literal_whitespace")) || (@token.type.space? && a_def.body.is_a?(Nop))
```
Comment (straight-shoota) [inline src/compiler/crystal/tools/formatter.cr:4284]: **suggestion:** Simplify the logic expression:

```suggestion
        write " " if a_def.body.is_a?(Nop) && (flag?("proc_literal_whitespace") || @token.type.space?)
```

**Review by straight-shoota — APPROVED** [review]

**Review by ysbaddaden — APPROVED** [review]
